### PR TITLE
Highlight spans with errors

### LINKED ISSUE #1140: For exception changing the Color to Red
Hi,
Can we have some feature where, if I have a tag added Key as "Exception" or  "error" the color changes to red , to have it highlighted

### PR DESCRIPTION
**Pull request by virtuald**
- Annotations: value must be 'error'
- Binary annotations: key must be 'error'
- Fixes #1140 

Here's three screenshots that illustrate the changes:

Traces view (top is transient, bottom is critical):
traces

Trace view (spans are critical, normal, critical + error annotation, transient):
trace

Trace modal:
trace-modal

And the JSON to duplicate:

```
[
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "error",
                "value": "You done messed up son"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7f42",
        "name": "get",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7e30",
        "parentId": "8e5b31230ccb7f42",
        "name": "getit",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "error",
                "value": "You done messed up son"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7e43",
        "parentId": "8e5b31230ccb7e30",
        "name": "gettt",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },{
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 19000,
        "id": "8e5b31230ccb7e53",
        "parentId": "8e5b31230ccb7e43",
        "name": "geterdone",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },{
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 19000,
        "id": "8e5b31230ccb7e99",
        "name": "is-transient",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7e99"
    }
]
```

### CODE REVIEW

File: zipkin-ui/js/component_ui/traceSummary.js [hunk 1]
```diff
@@ -124,6 +124,13 @@ function getSpanTimestamps(spans) {
   }))).value();
 }
 
+function detectErrors(spans) {
```
Comment (codefromthecrypt) [inline zipkin-ui/js/component_ui/traceSummary.js]: possible add a unit test for this?

File: zipkin-ui/js/component_ui/traceToMustache.js [hunk 2]
```diff
@@ -106,8 +106,12 @@ export default function traceToMustache(trace) {
       const spanStartTs = span.timestamp || traceTimestamp;
       const spanDepth = spanDepths[span.id] || 1;
       const width = (span.duration || 0) / summary.duration * 100;
+      let isError = false;
```
Comment (codefromthecrypt) [inline zipkin-ui/js/component_ui/traceToMustache.js]: not sure, but I _think_ this function is unit tested, if accessible, mind adding a test for 'error' -> isError?

**Comment by codefromthecrypt** [comment]
if you can, please add ERROR to this and use the constant. we can change the java, thrift sides in a different PR.

**Comment by codefromthecrypt** [comment]
this :) https://github.com/openzipkin/zipkin/blob/master/zipkin-ui/js/component_ui/traceConstants.js

File: zipkin-ui/js/component_ui/traceToMustache.js (hunk 2 above)
Comment (virtuald) [inline zipkin-ui/js/component_ui/traceToMustache.js]: I'll look into it tomorrow.

Comment (eirslett) [inline zipkin-ui/js/component_ui/traceToMustache.js]: You can do const isError = a.key === 'error' to avoid the let.

Comment (virtuald) [inline zipkin-ui/js/component_ui/traceToMustache.js]: @eirslett but a.key is inside of a closure, I don't think you will be able to reassign the const each time.

Comment (eirslett) [inline zipkin-ui/js/component_ui/traceToMustache.js]: Oh, I missed that, sorry!

**Comment by codefromthecrypt** [comment]
something to consider.. if we think of an "error" event as/annotation transient, we might want to stain the span yellow instead. Presence of the binary annotation could imply complete failure (something the client would be responsible for in an RPC span, or the server is responsible for in a server-only root span).

See https://github.com/openzipkin/zipkin-api/pull/18/files/1dc521c1994c104211f9a72400e48eae5e948549#r68687632 from @yurishkuro

**Comment by codefromthecrypt** [comment]
for example, I think finagle logs "finagle.retry" which is a stack-specific way of saying there was a transient error

**Comment by virtuald** [comment]
I'll try to get to the adjustments tomorrow, no time tonight.

**Comment by codefromthecrypt** [comment]
all your help has been appreciated, Dustin!

**Comment by jcarres-mdsol** [comment]
There should be some documentation so client implementors know that this functionality is available

**Comment by virtuald** [comment]
Updated code to include concept of transient (has annotation) vs critical (has binary annotation) errors, added unit tests, updated screenshots, updated JSON.

**Comment by codefromthecrypt** [comment]
great. will try it out soon!

**Comment by codefromthecrypt** [comment]
tried and looks good. thanks!

**Pull request merged**